### PR TITLE
[Refactor][Reduction] replace pytorch scalar fallback with native 0-D path

### LINKED ISSUE #1448: [REFACTOR][REDUCTION] remove pytorch scalar fallback
### Symptom / Motivation
Scalar reduction inputs currently short-circuit to PyTorch reference helpers in the Op layer. That keeps the API behavior aligned, but it makes the runtime implementation depend on an external library instead of TileOps/TileLang logic.

### Root Cause Analysis
- `tileops/ops/reduction/reduce.py` routes 0-D inputs through `_maybe_scalar()` and `_torch_scalar_reference()` for supported reductions.
- `docs/design/trust-model.md` and `manifest-spec.md` define PyTorch as the spec oracle, but the current scalar path treats it as a runtime fallback.
- The current design leaves no TileOps-native scalar path for the supported reduction family unless the scalar case is explicitly implemented or the limitation is documented as a true TileLang constraint.

### Related Files
- `tileops/ops/reduction/reduce.py`
- `.claude/domain-rules/manifest-spec.md`
- `docs/design/trust-model.md`
- `tests/ops/test_reduce_scalar_conformance.py`

## Goal
Replace the PyTorch runtime scalar fallback with a TileOps-native implementation or a documented, narrowly scoped TileLang exception, so reduction ops remain a TileOps runtime feature rather than delegating semantics back to PyTorch.

## Plan

**Plan type: proposal**
1. Audit every reduction path that currently uses `_maybe_scalar()` and decide whether TileOps can cover the 0-D case directly.
2. Implement a TileOps-native scalar path where feasible; keep any host fallback only for cases that are provably blocked by TileLang limitations.
3. Update the design / manifest language so PyTorch is clearly the spec oracle, not the default runtime implementation path.
4. Extend or adjust tests so scalar behavior and any documented exception boundary remain covered.

## Constraints
- Joint change across implementation and docs: runtime policy and its explanation should move together, not drift independently.
- Use PyTorch only as a correctness oracle in tests unless a documented TileLang limitation makes a runtime fallback unavoidable.
- Keep the change limited to reduction runtime behavior and its policy text; do not broaden it into unrelated op families.

## Acceptance Criteria
- [ ] AC-1: Modified files pass unit tests.
- [ ] AC-2: Scalar reduction paths no longer call PyTorch reference helpers at runtime where TileOps can implement the behavior directly.
- [ ] AC-3: The docs/specs clearly state when a host fallback is allowed and when TileOps must provide its own implementation.
- [ ] AC-4: Existing and new reduction tests cover scalar inputs and any documented exception boundary.

### PR DESCRIPTION
**Pull request by lcy-seso**
Closes #1448

## Summary

- Replace the PyTorch runtime scalar fallback in reduction ops with a TileOps-native 0-D path covering Sum/Mean/Amax/Amin/Prod/All/Any/CountNonzero/Std/Var/VarMean (and Welford variants).
- Update `docs/design/trust-model.md` and `.claude/domain-rules/manifest-spec.md` so PyTorch is documented as the spec oracle, not the default runtime implementation; allowed-fallback conditions are narrowed and named.
- Extend `tests/ops/test_reduce_scalar_conformance.py` and `tests/ops/test_reduction_scalar_input.py` to cover scalar inputs across all reduction ops and the documented exception boundary (invalid DOF).

## Test plan

- [x] AC-1: Modified files pass unit tests. — `CUDA_VISIBLE_DEVICES=3 python -m pytest -q tests/ops/test_reduction*.py tests/ops/test_reduce*.py tests/ops/test_*reduce*.py` → 1054 passed.
- [x] AC-2: Scalar reduction paths no longer call PyTorch reference helpers at runtime where TileOps can implement the behavior directly. — Helper-blocking probe patched `torch.sum/mean/amax/amin/prod/std/var/var_mean/all/any/count_nonzero` to raise; scalar ops still returned correct results.
- [x] AC-3: Docs/specs clearly state when a host fallback is allowed and when TileOps must provide its own implementation. — `docs/design/trust-model.md` and `.claude/domain-rules/manifest-spec.md` updated.
- [x] AC-4: Existing and new reduction tests cover scalar inputs and any documented exception boundary. — Scalar conformance and reduction scalar-input suites extended (+14 nodes).

## Test node delta

```
File                                           Base    HEAD    Delta
--------------------------------------------------------------------
tests/ops/test_reduce_scalar_conformance.py      96     108      +12
tests/ops/test_reduction_scalar_input.py         50      52       +2
--------------------------------------------------------------------
TOTAL                                           146     160      +14
```

**Justification:** Each new node pins a previously-uncovered scalar contract: prod / welford (var, std, var_mean) / logical (all, any) / count_nonzero scalar fast paths, plus the invalid-DOF exception boundary required by AC-4. No redundant coverage; the additions are the minimum needed to lock the new native 0-D path.

### CODE REVIEW

File: tileops/ops/reduction/reduce.py [hunk 1]
```diff
@@ -675,18 +672,32 @@ def _build_kernel_kwargs(self) -> dict:
...
+        if self.correction >= 1:
+            warnings.warn(
+                f"{self._op_kind}(): degrees of freedom is <= 0. Correction "
+                "should be strictly less than the reduction factor (input "
+                "numel divided by output numel).",
+                UserWarning,
+                stacklevel=2,
+            )
+            return torch.full((), float("nan"), dtype=x.dtype, device=x.device)
+        return torch.zeros((), dtype=x.dtype, device=x.device)
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/reduction/reduce.py]: Adopted. Welford scalar fast path now returns `x - x` so non-finite inputs propagate and autograd history on `x` is preserved. See b955011.

File: tileops/ops/reduction/reduce.py [hunk 2]
```diff
@@ -282,30 +278,31 @@ def _validate_scalar_dim(self) -> None:
...
+
+        Arithmetic reductions (``sum``, ``mean``, ``amin``, ``amax``,
+        ``prod``) over one element return the element itself. Logical /
+        count ops override ``_noop_output_dtype`` so this default applies
+        the ``x != 0`` predicate and casts to the declared output dtype.
+        Welford ops (``std``, ``var``, ``var_mean``) override this hook
+        because their result depends on ``correction``.
         """
-        return self._torch_scalar_reference(x)
+        out_dtype = self._noop_output_dtype()
```
Comment (Ibuki-wind) [inline tileops/ops/reduction/reduce.py:306]: This scalar branch now returns x.clone() / predicate output without ever checking whether dim aliases the same axis multiple times, so 0-D inputs like [0, -1] and [0, 0] silently succeed where PyTorch raises RuntimeError("dim 0 appears multiple times in the list of dims") -> reject duplicate canonical dims before returning here.

File: tileops/ops/reduction/reduce.py [hunk 3]
```diff
@@ -675,18 +672,34 @@ def _build_kernel_kwargs(self) -> dict:
...
-        if dim is None:
-            return ref(x, correction=self.correction)
-        return ref(x, dim=dim, correction=self.correction)
+        if self.correction >= 1:
+            warnings.warn(
+                f"{self._op_kind}(): degrees of freedom is <= 0. Correction "
+                "should be strictly less than the reduction factor (input "
+                "numel divided by output numel).",
+                UserWarning,
+                stacklevel=2,
```
Comment (Ibuki-wind) [inline tileops/ops/reduction/reduce.py:711]: _WelfordReduceOp._scalar_forward() builds fresh nan/0 tensors here, so 0-D var/std/var_mean with requires_grad=True lose the grad_fn and stop matching the PyTorch backward contract -> construct the closed form as an operation on x so the scalar output stays connected to the input.

**Review by Ibuki-wind — CHANGES_REQUESTED** [review]
### Overall
Blocked by two scalar-path regressions in `tileops/ops/reduction/reduce.py`; fix the native scalar validation and Welford autograd before merging.

File: tileops/ops/reduction/reduce.py (hunk 2 above)
Comment (lcy-seso) [reply tileops/ops/reduction/reduce.py:306]: Adopted. `_validate_scalar_dim` now canonicalizes dims on a 0-D tensor (both `0` and `-1` collapse to axis 0) and raises `RuntimeError('dim 0 appears multiple times in the list of dims')` when duplicates appear (`[0, 0]`, `[0, -1]`, etc.). See a0c6c64.

File: tileops/ops/reduction/reduce.py (hunk 3 above)
Comment (lcy-seso) [reply tileops/ops/reduction/reduce.py:711]: Adopted. `_WelfordReduceOp._scalar_forward` now builds the NaN branch as `x * float('nan')` (the zero branch was already `x - x`), so 0-D var/std/var_mean with `requires_grad=True` keep a grad_fn and match PyTorch's backward contract. See a0c6c64.

File: tileops/ops/reduction/reduce.py [hunk 4]
```diff
@@ -274,38 +270,46 @@ def _validate_scalar_dim(self) -> None:
                 )
             return
         if isinstance(dim, (list, tuple)):
```
Comment (Ibuki-wind) [inline tileops/ops/reduction/reduce.py:272]: _validate_scalar_dim() still only checks that each entry is 0/-1, so aliasing sequences like [0, -1] or [0, 0] continue into the scalar fast path instead of matching PyTorch's duplicate-dim error. -> Canonicalize the scalar dim sequence to axis 0 and reject duplicates before returning from _validate_scalar_dim().

File: tileops/ops/reduction/reduce.py [hunk 5]
```diff
@@ -675,18 +679,39 @@ def _build_kernel_kwargs(self) -> dict:
...
+        - ``N - correction <= 0`` (i.e. ``correction >= 1``): variance
+          and standard deviation are mathematically undefined; the
+          contract returns ``nan`` (computed as ``x * nan`` so the
+          output stays connected to ``x``).
+        - ``correction == 0``: the unbiased denominator is ``N``, so the
+          deviation from the mean (which equals the element itself) is
+          zero for finite inputs. The result is computed as ``x - x``
+          so non-finite inputs propagate (``nan`` / ``inf`` → ``nan``)
+          and autograd history on ``x`` is preserved.
+
```
Comment (Ibuki-wind) [inline tileops/ops/reduction/reduce.py:701]: This NaN branch still returns a fresh tensor for correction >= 1, so 0-D var/std/var_mean with requires_grad=True remain detached even though PyTorch's scalar var has a grad_fn here. -> Build the invalid-DOF NaN from tensor ops on x so the scalar output stays connected to autograd history.

**Review by Ibuki-wind — CHANGES_REQUESTED** [review]
### Overall
Two scalar-path conformance regressions remain: duplicate dim aliases still bypass the native 0-D contract, and the invalid-DOF Welford NaN path still detaches autograd history.

File: tileops/ops/reduction/reduce.py (hunk 4 above)
Comment (lcy-seso) [reply tileops/ops/reduction/reduce.py:272]: Rejecting — already in HEAD (a0c6c64). `_validate_scalar_dim` canonicalizes each list/tuple entry to axis 0 (both `0` and `-1` collapse) and tracks them in a `seen` set; the second occurrence of `canon=0` raises `RuntimeError('dim 0 appears multiple times in the list of dims')`. `[0, 0]` and `[0, -1]` both trigger the duplicate path. The review trailer references the prior commit (`b955011`); this fix landed in `a0c6c64`.

File: tileops/ops/reduction/reduce.py (hunk 5 above)
Comment (lcy-seso) [reply tileops/ops/reduction/reduce.py:701]: Rejecting — already in HEAD (a0c6c64). The correction>=1 branch returns `x * float('nan')` (line 713), which is a tensor op on `x` and therefore preserves the autograd graph (`MulBackward`); 0-D var/std/var_mean with `requires_grad=True` keep their `grad_fn`. The correction==0 branch is `x - x`, same property. The review trailer references the prior commit (`b955011`); this fix landed in `a0c6c64`.

File: tileops/ops/reduction/reduce.py [hunk 6]
```diff
@@ -274,38 +270,46 @@ def _validate_scalar_dim(self) -> None:
...
             return
         if isinstance(dim, (list, tuple)):
+            seen: set = set()
             for d in dim:
                 if d not in (0, -1):
                     raise IndexError(
                         f"Dimension out of range (expected to be in range of "
                         f"[-1, 0], but got {d})"
                     )
+                canon = 0  # 0 and -1 alias the same axis on a 0-D tensor.
```
Comment (Ibuki-wind) [inline tileops/ops/reduction/reduce.py:280]: _validate_scalar_dim() adds a new duplicate-dim branch, but no scalar test reaches it; existing scalar conformance only covers None, 0, -1, (), and [], so [0, -1]/[0, 0] can regress again. -> Add a scalar regression test that compares the duplicate-dim error against PyTorch for aliasing 0-D dims.

File: tileops/ops/reduction/reduce.py [hunk 7]
```diff
@@ -675,18 +679,39 @@ def _build_kernel_kwargs(self) -> dict:
...
-        return ref(x, dim=dim, correction=self.correction)
+        if self.correction >= 1:
+            warnings.warn(
+                f"{self._op_kind}(): degrees of freedom is <= 0. Correction "
+                "should be strictly less than the reduction factor (input "
+                "numel divided by output numel).",
+                UserWarning,
+                stacklevel=2,
+            )
+            return x * float("nan")
```
Comment (Ibuki-wind) [inline tileops/ops/reduction/reduce.py:713]: _WelfordReduceOp._scalar_forward() now preserves autograd with x * nan, but the scalar tests still assert only values/shapes, so the fixed requires_grad contract is untested. -> Add a 0-D Welford regression test with requires_grad=True that verifies the invalid-DOF output has autograd history/backward behavior matching PyTorch.

**Review by Ibuki-wind — CHANGES_REQUESTED** [review]
### Overall
Implementation fixes are in place, but the new scalar behaviors still need regression coverage before merge.

File: tileops/ops/reduction/reduce.py (hunk 6 above)
Comment (lcy-seso) [reply tileops/ops/reduction/reduce.py:280]: Adopted. Added `test_sum_scalar_duplicate_dim_matches_torch` covering `[0, 0]`, `[0, -1]`, `[-1, -1]`, `[-1, 0]`, `(0, -1)` — each parametrization asserts both PyTorch and TileOps raise `RuntimeError` with "appears multiple times". See 5e56655.

File: tileops/ops/reduction/reduce.py (hunk 7 above)
Comment (lcy-seso) [reply tileops/ops/reduction/reduce.py:713]: Adopted. Added `test_var_scalar_requires_grad_preserves_grad_fn` and `test_var_mean_scalar_requires_grad_preserves_grad_fn` — each builds a 0-D input with `requires_grad=True`, runs the op on the invalid-DOF (`correction=1`) NaN path, and asserts `requires_grad` + `grad_fn is not None` on every output, matching PyTorch. See 5e56655.

File: tileops/ops/reduction/reduce.py (hunk 4 above)
Comment (Ibuki-wind) [inline tileops/ops/reduction/reduce.py:272]: The duplicate-dim fast path is now correct in code, but there is still no scalar test that reaches it, so aliasing forms like [0, -1] and [0, 0] can regress silently. -> Add a 0-D regression test that asserts PyTorch's duplicate-dim RuntimeError for aliasing dims.

File: tileops/ops/reduction/reduce.py [hunk 8]
```diff
@@ -675,18 +679,39 @@ def _build_kernel_kwargs(self) -> dict:
...
+
+        ``VarMeanFwdOp`` overrides this hook to additionally return the
+        mean (the input element).
         """
-        ref = _TORCH_SCALAR_REF[self._op_kind]
-        dim = self.dim
-        if dim is None:
-            return ref(x, correction=self.correction)
-        return ref(x, dim=dim, correction=self.correction)
+        if self.correction >= 1:
```
Comment (Ibuki-wind) [inline tileops/ops/reduction/reduce.py:705]: The invalid-DOF NaN branch now preserves autograd in code, but the scalar suite still only checks values/shapes, so the requires_grad contract is untested. -> Add a 0-D Welford regression test with requires_grad=True that verifies grad_fn/backward behavior against PyTorch.

**Review by Ibuki-wind — CHANGES_REQUESTED** [review]
Divergence introspection: reaffirmed — `docs/design/trust-model.md` makes Test own correctness verification, and the new 0-D scalar branches still have no regression tests.
### Overall
The implementation fixes are in place, but the scalar 0-D paths still need regression tests before approval.

File: tileops/ops/reduction/reduce.py (hunk 4 above)
Comment (lcy-seso) [reply tileops/ops/reduction/reduce.py:272]: Rejecting — already in HEAD (5e56655). `tests/ops/test_reduction_scalar_input.py::test_sum_scalar_duplicate_dim_matches_torch` (added in 5e56655) parametrizes `[0, 0]`, `[0, -1]`, `[-1, -1]`, `[-1, 0]`, `(0, -1)` and asserts both `torch.sum` and `SumFwdOp` raise `RuntimeError` matching "appears multiple times". The thread anchor is 5e56655, which already contains the test.

File: tileops/ops/reduction/reduce.py (hunk 8 above)
Comment (lcy-seso) [reply tileops/ops/reduction/reduce.py:705]: Rejecting — already in HEAD (5e56655). `tests/ops/test_reduction_scalar_input.py::test_var_scalar_requires_grad_preserves_grad_fn` and `::test_var_mean_scalar_requires_grad_preserves_grad_fn` (added in 5e56655) build a 0-D input with `requires_grad=True`, run on the invalid-DOF (`correction=1`) NaN path, and assert `requires_grad` + `grad_fn is not None` on every output. The thread anchor is 5e56655, which already contains the tests.

**Review by Ibuki-wind — APPROVED** [review]
Divergence introspection: withdrawn — the prior blocker was test coverage, and the new tests now exercise the duplicate-dim and Welford scalar paths.
Clean — no issues.
